hail/src/test/scala/is/hail/expr/ir/IRSuite.scala

```diff
@@ -742,6 +740,459 @@ class IRSuite extends HailSuite {
     assertEvalsTo(MakeArray(FastSeq(), TArray(TInt32())), FastIndexedSeq())
   }
 
+  @Test def testMakeArrayInferPType() {
+    var ir = MakeArray(FastSeq(I32(5), NA(TInt32()), I32(-3)), TArray(TInt32()))
+
+    assertPType(ir, PArray(PInt32(false), true))
+
+    ir = MakeArray(FastSeq(I32(5), I32(-3)), TArray(TInt32()))
+    assertPType(ir, PArray(PInt32(true), true))
+
+    ir = MakeArray(
+          FastSeq(
+            MakeArray(FastSeq(I32(5), I32(5), I32(-3)), TArray(TInt32())),
+            MakeArray(FastSeq(I32(5), NA(TInt32()), I32(-3)), TArray(TInt32())),
+            MakeArray(FastSeq(I32(5)), TArray(TInt32()))
+          ),
+          TArray(TArray(TInt32()))
+        )
+
+    assertPType(ir, PArray(PArray(PInt32(), true), true))
+
+    ir = MakeArray(
+      FastSeq(
+        MakeArray(FastSeq(
+          MakeArray(FastSeq(I32(5), I32(-3), I32(-3)), TArray(TInt32())),
```

Comment:
these tests are pretty hard to read (and therefore maintain). The biggest barrier to nice tests seems to be that we don't have a way to easily specify a value of a given physical type.

The `In` node is the easiest way to do this. You can make that node take a physical type instead of a virtual type, but add an apply constructor so that we can change usages in a separate PR:
```
object In {
  def apply(t: Type): In = In(t.physicalType)
}
``` 

That way, your tests of MakeArray can be extremely terse and targeted, not relying on the physical types inferred by other nodes or recursively inferred by MakeArray children.